### ISSUE TITLE
Upload SBOM directly to OCI registry

### ISSUE THREAD

**Issue by developer-guy** [anchore/syft#592]
**What would you like to be added**:

cosign supports attaching SBOMs to OCI registries[^1] (also has a spec for it [^2]), so we (w/@dentrax) thought that it would be nice to have the same one for Syft, here is the design that we thought.

```shell
$ syft packages <image> -o oci://<repository>/<image>
```

**Why is this needed**:

**Additional context**:
[^1]: https://github.com/sigstore/cosign/blob/1f67ea73113f76c62429aa605d75105d77c370fa/cmd/cosign/cli/attach.go#L57
[^2]: https://github.com/sigstore/cosign/blob/main/specs/SBOM_SPEC.md

cc: @dlorenc @luhring

**Comment by developer-guy** [comment]
yeah, SGTM, we (w/@dentrax) can make it happen if you all agree on this `attach` command. We can start working on it ASAP. 

Let me summarize the steps that we are going to take:

1. Create a new command called `attach`.
2. Take a container image URL as an argument.
3. Generate an SBOM file for an image, and upload it to the OCI registry (maybe using cosign under the hood)

is that all we need @luhring, WDYT?

**Comment by luhring** [comment]
@developer-guy I think so! If we think of any more considerations, we'll put them here 😃 

Thanks for taking this on. 🙏 

Reach out at any point if it's not obvious to how use the existing patterns or Syft API for what you're doing and we'll jump in to take a look ❤️

**Comment by Dentrax** [comment]
> My two cents: a similar approach would make sense for Syft following the `cosign attach sbom ...` process:
> 
> ```
> syft attach <image>
> ```

I think making `attach` as a separated command would be making logic a bit hard to implement since `packagesExec()` function in `packages.go` file requires an `image` as an input argument. (We expect an SBOM to be generated before pass into cosign/attach command.)

The better UX IMHO would be passing a `--attach` flag in `packages` command. By this design, it'd be easier to call cosign/attach (if bool flag passed) just after `bus.Publish` in `packagesExecWorker()` func.

By the way, a new `attach` command could be useful in case we want to pass a _pre-generated_ SBOM file. So we would not be necessary to generate SBOM in the `attach` command.

cc: @developer-guy

**Comment by luhring** [comment]
>(We expect an SBOM to be generated before pass into cosign/attach command.)

Absolutely. On the implementation side, we can generate an SBOM in multiple execution paths. So we could make sure the `attach` command has an SBOM created before we invoke cosign's "attach" functionality, which should be all we need to do.

Another reason why `attach` makes sense as a new command is that it has its own distinct configuration needs, such as "write" access to registries (Syft's other commands don't require this), whether or not to override an existing attachment, etc.

I think the `attest` and `attach` functions should probably look the same on the command line, meaning they should either both be subcommands or they should both be flags, or else it might be confusing to users. IMHO, choosing to implement these as subcommands is more inline with the direction Syft is heading (see #516). cc: @wagoodman and @spiffcs for fact-checking me on this! 😄 

>By the way, a new attach command could be useful in case we want to pass a pre-generated SBOM file. So we would not be necessary to generate SBOM in the attach command.

I'm not sure I understand this part. If an SBOM has already been generated, why would users want to do the "attach" with `syft` instead of just using `cosign attach`?

**Comment by samj1912** [comment]
This was discussed in the Anchore community WG meeting today (notes available at https://docs.google.com/document/d/1ZtSAa6fj2a6KRWviTn3WoJm09edvrNUp4Iz_dOjjyY8/edit ) - 

These were the options that were proposed - 

1. We could implement this in syft
   - it could potentially leverage the multi-output syntax being introduced by https://github.com/anchore/syft/pull/732 to have something like `-o json=registry://my/image:latest`
   - syft could potentially import cosign as a library using the logic at https://github.com/sigstore/cosign/blob/f19f4f754de43e8151b941ed2307374dae55381d/cmd/cosign/cli/attach/sbom.go#L34-L61
2. We could continue to let cosign handle SBOM/attachment
   - we could explicitly document how the two tools interact in the syft docs (or even in the cosign docs?) so as to let users know that they can work with each other
   - Since cosign appects sbom via stdin, this would allow syft to directly upload an sbom to the registry via `syft packages <image> -o json | cosign attach sbom <image> --sbom - --type syft`. This also follows the unix philosophy of one tool doing one thing well, while being able to be chained with other tools

In the WG meeting, people were leaning towards the latter + clear documentation to note that syft is compatible with cosign unless there are some other advantages to natively supporting cosign in syft.

**Comment by dlorenc** [comment]
The cosign behavior is documented here as a "spec", with the intention of being interoperable!

https://github.com/sigstore/cosign/blob/main/specs/SBOM_SPEC.md

Either approach sounds great to me.

**Comment by samj1912** [comment]
Added https://github.com/sigstore/cosign/pull/1278 to update the spec as well

**Comment by luhring** [comment]
Just catching up — if we're going with option 2 from the [comment](https://github.com/anchore/syft/issues/592#issuecomment-1006821718) above, does that mean there's no work in Syft code for this any more, and it's just documentation? (of the flow for `syft ... | cosign attach sbom ...`?)

**Comment by samj1912** [comment]
@luhring that's correct :) https://github.com/sigstore/cosign/pull/1137 and https://github.com/sigstore/cosign/pull/1278 handle the syft support and documentation on the cosign side. We just need to do the same at https://github.com/anchore/syft#adding-an-sbom-to-an-image-as-an-attestation to also mention sbom attachments

**Comment by luhring** [comment]
Perfect, thanks @samj1912!

**Comment by developer-guy** [comment]
Hi folk, as @samj1912 said, syft is now capable of uploading SBOM in form of in-toto attestation format to the registries, also, there is another option like "cosign attach", but AFAIK, there is still no direct support for uploading SBOMs to the OCI registries. Maybe, we could add another output option like "oci" to support this, the UX would be:

```shell
# will upload SBOM in a json format by using "application/vnd.syft+json" mediaType
$ syft packages alpine:latest -o json --upload/export oci://devopps/my-sbom-img 
```

@dentrax @luhring

**Comment by developer-guy** [comment]
kindly ping

**Comment by kzantow** [comment]
From discussion:
* using `oci` could mean a number of different things, so we should tread carefully with naming, for example does this mean `OCI Reference Types` or `OCI Attachments`?

**Comment by xnox** [comment]
I hope all the relevant people are still here. I am very new to this, and want to learn prior art and what are the up to date current routes are possible going forward.

Over at cosign docs it says:
> WARNING: SBOM attachments are deprecated and support will be removed in a Cosign release soon after 2024-02-22 (see https://github.com/sigstore/cosign/issues/2755). Instead, please use SBOM [attestations](https://github.com/sigstore/cosign/blob/main/specs/ATTESTATION_SPEC.md).

Thus is this correct that the syft media type in that spec, is also deprecated? (the `application/vnd.syft+json`). Or can it and should it still be used in the attestations? That spec says

> This document assumes you are using the In-Toto [Attestation](https://github.com/in-toto/attestation) format, serialized as a DSSE envelope Other formats can be used, and the mediaType property should describe the format of a particular attestation, but implementations may not understand them. The DSSE envelope format is defined [here](https://github.com/secure-systems-lab/dsse/blob/master/envelope.md#dsse-envelope) and uses the mediaType: application/vnd.dsse.envelope.v1+json.

I am most familiar with intoto statements as attestations, most notably SPDX, and those documents have intoto predicates registered for example [v2.3](https://github.com/in-toto/attestation/blob/main/spec/predicates/spdx2.md) and [v3](https://github.com/in-toto/attestation/blob/main/spec/predicates/spdx3.md). Should in similar manner predicates registered for syft json i.e. something like `http://anchore.io/syft/json` ? (no need for version, as syft json declares schema versions which overall makes them stable. Although can also support  `http://anchore.io/syft/json/16.1.3` predicates straight away - such that multiple can be attached without needing to do a later transition like spdx is undergoing).

The idea is that if a container in an OCI registry has a cosign attestation that includes `http://anchore.io/syft/json` grype could pull that; and produce scan results based on that. If it fails to parse that attestation, it would fallback to pulling the full container and generating syft.json from scratch. Or if attestation is deemed too old and/or a different one could be generated - whichever mechanics grype currently does when passed in syft.json.

Does above make sense? Is it appropriate to attempt to register intoto syft.json predicate?

**Comment by sudo-bmitch** [comment]
A lot has changed in cosign since this was opened. Some of the current ideas in cosign include attaching artifacts with the OCI subject/referrers, and then signing those manifests the same as signing container images. https://github.com/sigstore/cosign/issues/3599

**Comment by kzantow** [comment]
Hi, I'm just trying to clarify what the ask is:
* use _mediaType_ `application/vnd.dsse.envelope.v1+json`
* add Syft [in-toto attestation](https://github.com/in-toto/attestation/blob/main/spec/predicates) of `http://anchore.io/syft/json` (or similar) without specifying a version?
* we should re-think everything due to introduction of `referrers`?

Any clarity here would be very helpful!
